frontend-react/docs/test-conventions.md

```diff
@@ -0,0 +1,492 @@
+        });
+    });
+})
+```
+
+> Why?
+>
+> Having clearer and more sensible descriptions calls attention to what the purpose of the test actually is, which makes it easier to follow along with the underlying logic or to group similarly intentioned tests together.
+
+
+### Tests should check behavior, not implementation
+
+Especially in the world of front-end code, it's especially important to test our code from a user's perspective -- this means our tests should be prioritizing checking behavior ("what happens") rather than implementation details ("how it happens").
+
+```tsx
+function LoginForm() {
+    function handleSubmit() {
+        // assume a POST is sent to create a user session
+    }
+    
+    return (
+        <form onSubmit={handleSubmit}>
+            <button type="submit">
+        </form>
+    );
+}
+```
+
+> Why?
+>
+> Testing behavior keeps tests agnostic to the implementation, which makes them more resistant to changes and refactors.  Ultimately, changing implementation shouldn't automatically imply needing to update tests unless the behavior is also changing.
+
+Somewhat of an exception to the aforementioned rule is pure functions, whose behavior is essentially its implementation:
+
+```tsx
+// source code
+function capitalize(str?: string) {
+    if (!str) {
+        return "";
+    }
+    
+    return str.charAt(0).toUpperCase() + str.substring(1).toLowerCase();
+}
+
+// tests
+describe('capitalizeString', () => {
+    describe("when the provided string is falsy", () => {
+        test("returns an empty string", () => {
+            expect(capitalizeString()).toEqual("");
+            expect(capitalizeString("")).toEqual("");
+        });
+    });
+
+    describe("when the provided string is truthy", () => {
+        test("returns the string with the first character capitalized (if applicable)", () => {
+            expect(capitalizeString("a")).toEqual("A");
+            expect(capitalizeString("A")).toEqual("A");
+			
+            expect(capitalizeString("abc")).toEqual("Abc");
+            expect(capitalizeString("Abc")).toEqual("Abc");
+            expect(capitalizeString("ABc")).toEqual("Abc");
+            expect(capitalizeString("ABC")).toEqual("Abc");
+    
+            expect(capitalizeString("1")).toEqual("1");
+            expect(capitalizeString("1abc")).toEqual("1abc");
+    
+            // and so on...
+        });
+    });
+});
+```
+
+### Tests should be as explicit as possible
+
+```tsx
+// bad
+expect(sum(1, 2, 3)).toEqual(1 + 2 + 3);
+
+const arrayA = [1, 2, 3];
+const arrayB = [4, 5, 6];
+const arrayC = [7, 8, 9];
+expect(
+    concatenate(arrayA, arrayB, arrayC)
+).toEqual([
+    arrayA,
+    arrayB,
+    arrayC
+].flat());
+       
+// good
+expect((sum(1, 2, 3))).toEqual(6);
+expect(
+    concatenate(arrayA, arrayB, arrayC)
+).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9]);
+```
+
+> Why?
+>
+> Tests should read as superficially clearly as possible, and they shouldn't reimplement the underlying logic.
```

Comment:
Should we be clearer here that the expectation is that the compare value, no matter how complex, needs to precalculated?

Reply:
Yeah, that's a good point!  I'd say the compare value doesn't necessarily have to be precalculated as long as it's clear (could be through variable names like `VALID_VALUE`/`TODAY`/`ONE_MILLION_SECONDS`) at first glance and doesn't require the reader of the test to trace through logic.